### ISSUE TITLE
ARM builds for Linux

### ISSUE THREAD

**Issue by alcuadrado** [ethereum/solidity#11351]
With Apple switching to ARMv8 releasing ARM builds of solc is increasingly important, especially linux ARM builds.

The reason for this is that if you use Docker with an ARM mac, you'll end up running your containers in an ARM-based Linux, so you can't run any native solc build.

In Mac the situation is different, you can still run x86_64 binaries using Rosetta2, and this process is 100% transparent.

Hardhat, and I guess all the other Node.js-based tools, automatically fallbacks to socjs, so it still works in ARM-based Docker containers. Other tools, like Brownie, probably break in that situation.

**Comment by axic** [comment]
We have set up ARM builds on another project, but it is rather computationally expensive as it is running in an emulator. I wonder how big of a slowdown that would be for running soltest within that emulator, because without running tests we can't be satisfied it actually works, which reduces its usefulness.

And as you said, macOS still works out of the box. We are waiting for CircleCI to release support, which they may do this year :sweat:

However I really wonder why would projects start doing an ARM Linux build, which is not really used outside of RaspberryPI PC builds? Why not just wait for proper macOS build environments, as that is the target people actually have? Do you know of an example project which already does this?

**Comment by alcuadrado** [comment]
The only reason I can think of is because of docker running on an M1 mac. It's really annoying, because switching to arm in a mac is pretty straightforward, except when using docker. 

There are probably other reasons though, as some projects, like Node.js were already distributing ARM linux builds.

**Comment by axic** [comment]
Are you saying Docker on an M1 mac virtualises ARM and is not running on top of Rosetta? Yeah that may be an annoying situation and a strong enough reason.

Though I personally only used docker once or twice on a mac because it was so dead slow compared to linux.

**Comment by alcuadrado** [comment]
> Are you saying Docker on an M1 mac virtualises ARM and is not running on top of Rosetta? 

Yes, exactly. Docker desktop runs an arm linux vm and the containers inside it.

> Though I personally only used docker once or twice on a mac because it was so dead slow compared to linux.

Yeah, it's annoying. I think the slow down is because of the vm.

**Comment by cameel** [comment]
This should be now possible. On our last call with people from CircleCI they had this in their presentation:

> **Key Releases Past 6 Months**
>
> - ARM Cloud Support
>     - Medium and large ARM machines are now a part of our cloud platform
>     - Documents here: https://circleci.com/docs/2.0/arm-resources/

[Arm Resources on CircleCI](https://circleci.com/changelog/#arm-resources-on-circleci)
> February 01, 2021
>
> We have initiated a preview of hosted Arm resources on CircleCI. Arm is an architecture that is quickly gaining traction for server workloads. Organizations choose Arm because of its faster performance, which results in faster workloads and significant cost savings compared to x86-based architectures. An essential piece of migrating workloads to Arm is having Arm support in your CI system. If you are on a usage-based plan, including the Free plan, you may request access to the Arm preview. We are making two arm64 resources available: arm.medium and arm.large.
>
> To request access, customers must fill out an access form that is linked in the preview docs below.

**Comment by dezzeus** [comment]
Any update ? @cameel

**Comment by cameel** [comment]
Sorry, nothing yet. We'll get to this eventually but in the short term there are a few things from our roadmap that have priority, like getting the IR compilation pipline ready for general use.

**Comment by gakonst** [comment]
FYI @nikitastupin went ahead and made a GHA release flow here https://github.com/nikitastupin/solc/. Would love exploring a way to get this upstreamed.

**Comment by cameel** [comment]
That looks nice. This is only for the latest version though, isn't it? I don't see any workarounds that would probably be required to build earlier versions. I mean, if none are needed then great, but I needed a ton when I did builds for macOS earlier (see how gnarly [`random-macosx-build.yml`](https://github.com/ethereum/solc-bin/blob/gh-pages/.github/workflows/random-macosx-build.yml) is :))

But anyway, if you're interested in upstreaming this, I don't think we'd accept binaries but if @nikitastupin is willing to refactor this script to work in a similar vein to [`random-macosx-build.yml`](https://github.com/ethereum/solc-bin/blob/gh-pages/.github/workflows/random-macosx-build.yml), we'd run it in our CI and build them ourselves.

One important thing is that the script must run our test suite on the built binary to verify it's built correctly. At least to the extent that these older versions allowed - the oldest ones didn't and later ones could run only part of the test suite without setting up a node. The action you linked to does not do this and this part is actually what added a lot of complexity in the macOS case. I think it's crucial though.

If you do this, you can submit the script in a PR to [solc-bin](https://github.com/ethereum/solc-bin/) and I'll help reviewing it. Here's my PR for macOS: https://github.com/ethereum/solc-bin/pull/53 for reference. See in particular https://github.com/ethereum/solc-bin/pull/53#issuecomment-700149493, which explains the specific workarounds needed - most of which hopefully will not be needed on Linux.

Once the binaries are built, I'll submit them in a separate PR, which will trigger the bytecode comparison check ([`t-bytecode-compare.yml`](https://github.com/ethereum/solc-bin/blob/gh-pages/.github/workflows/t-bytecode-compare.yml)). This will ensure that the bytecode produced by the binaries is identical with other platforms, at least to the extent covered by our tests.

---

This is only about rebuilding existing versions though. Another part of this task is to add the new platform to our release process. This is fairly straightforward and the blocker here was simply lack of support on CircleCI's part. If you want to submit a PR for that part, it would basically be this:
1) Add a CircleCI job similar to [`b_ubu_static`](https://github.com/ethereum/solidity/blob/develop/.circleci/config.yml#L788-L803) but running on ARM.
    - I think we'll also need equivalents of `t_ubu_soltest` and `t_ubu_cli` jobs for this platform.
2) Update [`ReleaseChecklist.md`](https://github.com/ethereum/solidity/blob/develop/ReleaseChecklist.md).

**Comment by axic** [comment]
It is possible now to use CircleCI's ARM machines, but the image they have is a rather old Ubuntu. We use it in [Fizzy](https://github.com/wasmx/fizzy/blob/master/circle.yml#L519-L522):
```yaml
  arm64-native:
    machine:
      image: ubuntu-2004:current
      resource_class: arm.large
    steps:
```

**Comment by axic** [comment]
Even windows+arm is now supported on github actions too: https://github.blog/changelog/2022-09-28-github-actions-self-hosted-runners-now-support-windows-arm-hardware/

**Comment by nikitastupin** [comment]
GitHub has [launched ARM-based runners](https://github.blog/news-insights/product-news/arm64-on-github-actions-powering-faster-more-efficient-build-systems/) a few months ago. It's only accessible to orgs on GitHub Team and Enterprise Cloud plans though.

**Comment by felix314159** [comment]
Please upgrade this from "Consider to promote" to "To do" in your focus tasks. EEST relies on solc-select which currently will gladly download an x86 binary on arm linux machines, leading to issues. Providing official arm binaries disincentivizes workarounds such as using tools like [svm-rs](https://github.com/alloy-rs/svm-rs/) which will work on arm linux but just grab pre-built binaries from [this person](https://github.com/nikitastupin/solc/tree/main/linux/aarch64). I am aware that the Ethereum org is currently trying to get a better GitHub plan so that this becomes possible via ARM runners, just know that 4 years after this thread has started there is an ever growing demand of official arm linux binaries. Ty!

**Comment by cameel** [comment]
Quick update: this is already mostly done in #16268, #16271 and https://github.com/argotorg/solc-bin/pull/167/. The only thing that is missing are soltest and CLI test runs using the new binaries.

0.8.31 tomorrow will be the first release that will include linux ARM binaries.

We are not planning to provide builds for older versions in the near future. We could consider that in the longer term, but in line with https://github.com/argotorg/solidity/issues/12291#issuecomment-3604441475 I'm not going to consider that a part of this issue.
